cluster/gce/windows/common.psm1

```diff
@@ -146,6 +146,307 @@ function MustDownload-File {
   }
 }
 
+# This compiles some C# code that can make syscalls, and pulls the
+# result into our powershell environment so we can make syscalls from this script.
+# We make syscalls directly, because whatever the powershell cmdlets do under the hood,
+# they can't seem to open the log files concurrently with writers.
+# See https://docs.microsoft.com/en-us/dotnet/framework/interop/marshaling-data-with-platform-invoke
+# for details on which unmanaged types map to managed types.
+$SyscallDefinitions = @'
+[DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+public static extern IntPtr CreateFileW(
+  String lpFileName,
+  UInt32 dwDesiredAccess,
+  UInt32 dwShareMode,
+  IntPtr lpSecurityAttributes,
+  UInt32 dwCreationDisposition,
+  UInt32 dwFlagsAndAttributes,
+  IntPtr hTemplateFile
+);
+
+[DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+public static extern bool SetFilePointer(
+  IntPtr hFile,
+  Int32  lDistanceToMove,
+  IntPtr lpDistanceToMoveHigh,
+  UInt32 dwMoveMethod
+);
+
+[DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+public static extern bool SetEndOfFile(
+  IntPtr hFile
+);
+
+[DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+public static extern bool CloseHandle(
+  IntPtr hObject
+);
+'@
+$Kernel32 = Add-Type -MemberDefinition $SyscallDefinitions -Name 'Kernel32' -Namespace 'Win32' -PassThru
+
+# ideally we can just use setFilePointer and setEndOfFile to truncate
+
+function Close-Handle {
+  param (
+    [parameter(Mandatory=$true)] [System.IntPtr]$Handle
+  )
+  $ret = $Kernel32::CloseHandle($Handle)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if (-not $ret) {
+    throw "Failed to close open file handle ${Handle}, system error code: ${err}"
+  }
+}
+
+function Open-File {
+  param (
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+
+  $lpFileName = $Path
+  $dwDesiredAccess = [System.IO.FileAccess]::ReadWrite
+  $dwShareMode = [System.IO.FileShare]::ReadWrite # Fortunately golang also passes these same flags when it creates the log files, so we can open it concurrently.
+  $lpSecurityAttributes = [System.IntPtr]::Zero
+  $dwCreationDisposition = [System.IO.FileMode]::Open
+  $dwFlagsAndAttributes = [System.IO.FileAttributes]::Normal
+  $hTemplateFile = [System.IntPtr]::Zero
+
+  $handle = $Kernel32::CreateFileW($lpFileName, $dwDesiredAccess, $dwShareMode, $lpSecurityAttributes, $dwCreationDisposition, $dwFlagsAndAttributes, $hTemplateFile)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if ($handle -eq -1) {
+    if ($handle -ne -1) {
```

Comment:
Looks like this can never be entered, based on previous line.

Reply:
Oh, good catch, thanks!

Reply:
Confirmed that the return value is always a valid handle when no error occurs. https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-createfilea#return-value